src/Orleans.Runtime/Catalog/Catalog.cs

```diff
@@ -1363,20 +1202,27 @@ public SiloStatus LocalSiloStatus
 
         public Task DeleteActivations(List<ActivationAddress> addresses)
         {
-            return DestroyActivations(TryGetActivationDatas(addresses));
-        }
-
+            List<ActivationData> TryGetActivationDatas(List<ActivationAddress> addresses)
+            {
+                var datas = new List<ActivationData>(addresses.Count);
+                foreach (var activationAddress in addresses)
+                {
+                    ActivationData data;
+                    if (TryGetActivationData(activationAddress.Activation, out data))
+                        datas.Add(data);
+                }
+                return datas;
+            }
 
-        private List<ActivationData> TryGetActivationDatas(List<ActivationAddress> addresses)
-        {
-            var datas = new List<ActivationData>(addresses.Count);
-            foreach (var activationAddress in addresses)
+            var timeoutTokenSource = new CancellationTokenSource(this.collectionOptions.Value.DeactivationTimeout);
+            var taskCompletionSourceList = new List<TaskCompletionSource<object>>(addresses.Count);
+            var counter = 0;
+            foreach (var activationData in TryGetActivationDatas(addresses))
             {
-                ActivationData data;
-                if (TryGetActivationData(activationAddress.Activation, out data))
-                    datas.Add(data);
+                DestroyActivationAsync(activationData, taskCompletionSourceList[counter], timeoutTokenSource.Token);
+                counter++;
             }
-            return datas;
+            return Task.WhenAll(taskCompletionSourceList.Select(tcs => tcs.Task));
```

Comment:
Why not make DestroyActivationAsync true async (`async Task`, not `async void`) and eliminate the TCS, waiting on the true task instance?

Also, wont `taskCompletionSourceList[counter]` throw an index out of bounds exception?